Add ADT variant infomation to StableMIR and finish implementing TyKind::internal()

celinval opened the pull request:
Introduce a `VariantDef` type and a mechanism to retrieve the definition from an `AdtDef`.

The `VariantDef` representation itself is just a combination of `AdtDef` and `VariantIdx`, which allow us to retrieve further information of a variant. I don't think we need to cache extra information for now, and we can translate on an on demand manner. I am  leaving the fields public today due to https://github.com/rust-lang/project-stable-mir/issues/56, but they shouldn't. For this PR, I've only added a method to retrieve the variant name, and its fields. I also added an implementation of `RustcInternal` that allow users to retrieve more information using Rust internal APIs.

I have also finished the implementation of `RustcInternal` for `TyKind` which fixes https://github.com/rust-lang/project-stable-mir/issues/46.

## Motivation

Both of these changes are needed in order to properly interpret things like projections. For example,
- The variant definition is used to find out which variant we are downcasting to.
- Being able to create `Ty` from `TyKind` helps for example processing each stage of a projection, like the code in `place.ty()`.

celinval committed:
Add Variant and a few more APIs to stable_mir

celinval committed:
Finish implementing `RustcInternal` for `TyKind`

This will allow us to provide methods to create `Ty` inside the stable
MIR, which can be helpful while handling pointers and other stuff.

bors commented:
:umbrella: The latest upstream changes (presumably #118543) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

kwfn commented on compiler/stable_mir/src/mir/body.rs:
```diff
@@ -492,12 +498,32 @@ pub struct Place {
     pub name: Symbol,
+
+    /// Source info of the user variable, including the scope
+    /// within which the variable is visible (to debuginfo)
```
```suggestion
    /// within which the variable is visible (to debuginfo).
```

ouz-a commented:
Will try to review this week.

celinval committed:
Add FieldDef to StableMIR and methods to get type

ouz-a commented on compiler/rustc_smir/src/rustc_internal/internal.rs:
```diff
@@ -297,6 +470,7 @@ where
         (*self).internal(tables)
     }
 }
+
```
nit

celinval replied:
Do you want me to remove this? I added a space here since these two implementations had no space in between.

ouz-a replied:
ah no problem then

ouz-a commented on compiler/rustc_smir/src/rustc_smir/context.rs:
```diff
@@ -240,6 +255,18 @@ impl<'tcx> Context for TablesWrapper<'tcx> {
+        let mut tables = self.0.borrow_mut();
+        let args = args.internal(&mut *tables);
+        let def_ty = tables.tcx.type_of(item.internal(&mut *tables));
+        // FIXME(celinval): use try_fold instead to avoid crashing.
```
How does this crash occur, any way to trigger it, if so can we solve it now ?

celinval replied:
I did some investigation. I couldn't find any easy fix, such as an internal API that we could use here. I've tried using `try_instantiate_and_normalize_erasing_regions`, but the first thing it does is to invoke `instantiate()` so the same few crashes still happen. It's looking like we would either have to implement one ourselves, e.g., implement a `FallibleTypeFolder`, or extend the `EarlyBinder` to expose a `try_instantiate` method.

That said, the only two scenarios I manage to trigger a crash were:
1. The arguments length is less than the number of generic arguments.
2. There's an unexpected `GenericArgumentKind`.

From an API perspective, I would like to avoid crashes whenever possible, and that's why I wanted to make this API fallible. That said, it's looking pretty hard to do proper validation. I do wonder if we should remove the `Result<>` from this API. I think many of our APIs will unfortunately depend on the user doing the correct thing since at the MIR level we are working on, we expect types to have been checked already.

ouz-a replied:
Yeah your points are valid, this isn't easy to re-produce my main concern with this was if it was easy to re-produce or easy to crash. People could easy crash the compiler if they play with it long enough this shouldn't block this PR from passing.

ouz-a reviewed (changes requested):
Overall looks very good, I'm happy that you provided very detailed documentation, and have a clear goal with this PR. just left a nit and a comment.

celinval committed:
Change ty_with_args to return Ty instead of Result

Although, we would like to avoid crashes whenever
possible, and that's why I wanted to make this API fallible. It's
looking pretty hard to do proper validation.

I think many of our APIs will unfortunately depend on the user doing
the correct thing since at the MIR level we are working on,
we expect types to have been checked already.

ouz-a commented:
r=me when green

celinval commented:
@bors r=@ouz-a

@bors rollup

bors commented:
:pushpin: Commit 326fea0fb86e148adc8168ac70a5ceddef3ad9a3 has been approved by `ouz-a`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ouz-a 326fea0fb86e148adc8168ac70a5ceddef3ad9a3 -->
<!-- homu: {"type":"Approved","sha":"326fea0fb86e148adc8168ac70a5ceddef3ad9a3","approver":"ouz-a","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
